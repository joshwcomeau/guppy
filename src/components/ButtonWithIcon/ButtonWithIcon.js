// @flow
import React from 'react';
import styled from 'styled-components';

import { StrokeButton } from '../Button';

import type { StrokeButtonProps } from '../Button';

type Props = StrokeButtonProps & {
  icon: React$Node,
};

// Note: noPadding prop seems unavailable on StrokeButton component but
//       it is available as it is passed down with delegated props to ButtonBase component.
const ButtonWithIcon = ({ icon, children, ...delegated }: Props) => (
  // TODO: Support other sizes
  <StrokeButton noPadding {...delegated}>
    <InnerWrapper>
      <IconWrapper>{icon}</IconWrapper>
      {children}
    </InnerWrapper>
  </StrokeButton>
);

export const InnerWrapper = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  height: 34px;
  padding: 0 10px 0 32px;
`;

const IconWrapper = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  width: 34px;
  height: 34px;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export default ButtonWithIcon;
