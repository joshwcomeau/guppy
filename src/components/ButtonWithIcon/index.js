// @flow
export { default } from './ButtonWithIcon';
