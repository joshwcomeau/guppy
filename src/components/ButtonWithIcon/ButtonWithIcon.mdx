---
name: ButtonWithIcon
menu: Reusable Components
---

import { Playground, Props } from 'docz';
import styled from 'styled-components';
import IconBase from 'react-icons-kit';
import ButtonWithIcon from './ButtonWithIcon';
import { settings } from 'react-icons-kit/feather';
import reactIconSrc from '../../assets/images/react-icon.svg';

# ButtonWithIcon

<Props of={ButtonWithIcon} />

## Basic usage

### React
<Playground>
  <ButtonWithIcon
    icon={<img src={reactIconSrc} style={{ width: '32px', height: '32px'}} />}
  >
    Check Button
  </ButtonWithIcon>
</Playground>

### Settings
<Playground>
  <ButtonWithIcon
    icon={<IconBase icon={settings} />}
  >
    Settings
  </ButtonWithIcon>
</Playground>
