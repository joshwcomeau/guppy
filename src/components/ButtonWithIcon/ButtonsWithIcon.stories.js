// @flow
import React from 'react';
import styled from 'styled-components';
import IconBase from 'react-icons-kit';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';
import { check, settings } from 'react-icons-kit/feather';

import reactIconSrc from '../../assets/images/react-icon.svg';

import Showcase from '../../../.storybook/components/Showcase';
import Paragraph from '../Paragraph';
import ButtonWithIcon from './ButtonWithIcon';

storiesOf('ButtonWithIcon', module).add(
  'default',
  withInfo()(() => (
    <React.Fragment>
      <Showcase label="React">
        <ButtonWithIcon
          icon={<ReactIcon src={reactIconSrc} />}
          onClick={action('clicked')}
        >
          Check Button
        </ButtonWithIcon>
      </Showcase>
      <Showcase label="Settings">
        <ButtonWithIcon
          icon={<IconBase icon={settings} />}
          onClick={action('clicked')}
        >
          Settings
        </ButtonWithIcon>
      </Showcase>
    </React.Fragment>
  ))
);

const ReactIcon = styled.img`
  width: 32px;
  height: 32px;
`;
