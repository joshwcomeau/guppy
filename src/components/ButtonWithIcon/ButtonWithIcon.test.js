/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { mount } from 'enzyme';
import ButtonWithIcon, { InnerWrapper } from './ButtonWithIcon';

describe('ButtonWithIcon component', () => {
  it('should render button with icon', () => {
    // mock icon to reduce snapshot size
    const icon = <svg />;
    const wrapper = mount(<ButtonWithIcon icon={icon} />);
    // Snapshot of InnerWrapper is OK as StrokeButton is tested separately.
    expect(wrapper.find(InnerWrapper)).toMatchSnapshot();
  });
});
