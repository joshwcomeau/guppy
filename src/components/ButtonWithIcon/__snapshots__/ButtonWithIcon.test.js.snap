// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonWithIcon component should render button with icon 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 34px;
  padding: 0 10px 0 32px;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  width: 34px;
  height: 34px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<styled.div>
  <div
    className="c0"
  >
    <styled.div>
      <div
        className="c1"
      >
        <svg />
      </div>
    </styled.div>
  </div>
</styled.div>
`;
